torchvision/prototype/datasets/_builtin/food101.py

```diff
@@ -0,0 +1,97 @@
+from pathlib import Path
+from typing import Any, Tuple, List, Dict, Optional, BinaryIO
+
+from torchdata.datapipes.iter import (
+    IterDataPipe,
+    Filter,
+    Mapper,
+    LineReader,
+    JsonParser,
+    Demultiplexer,
+    UnBatcher,
+    IterKeyZipper,
+)
+from torchvision.prototype.datasets.utils import Dataset, DatasetInfo, DatasetConfig, HttpResource, OnlineResource
+from torchvision.prototype.datasets.utils._internal import (
+    MappingIterator,
+    hint_shuffling,
+    hint_sharding,
+    path_comparator,
+    getitem,
+    INFINITE_BUFFER_SIZE,
+)
+from torchvision.prototype.features import Label, EncodedImage
+
+
+class Food101(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "food101",
+            homepage="https://data.vision.ee.ethz.ch/cvl/datasets_extra/food-101",
+            valid_options=dict(split=("train", "test")),
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        return [
+            HttpResource(
+                url="http://data.vision.ee.ethz.ch/cvl/food-101.tar.gz",
+                sha256="d97d15e438b7f4498f96086a4f7e2fa42a32f2712e87d3295441b2b6314053a4",
+                preprocess="decompress",
+            )
+        ]
+
+    def _classify_archive(self, data: Tuple[str, Any]) -> Optional[int]:
+        path = Path(data[0])
+        if path.parents[1].name == "images":
+            return 0
+        elif path.parents[0].name == "meta":
+            return 1
+        else:
+            return None
+
+    def _prepare_sample(self, data: Tuple[str, Tuple[str, BinaryIO]]) -> Dict[str, Any]:
+        id, (path, buffer) = data
+        return dict(
+            label=Label.from_category(id.split("/", 1)[0], categories=self.categories),
+            path=path,
+            image=EncodedImage.from_file(buffer),
+        )
+
+    def _image_key(self, data: Tuple[str, Any]) -> str:
+        path = Path(data[0])
+        return path.relative_to(path.parents[1]).with_suffix("").as_posix()
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+    ) -> IterDataPipe[Dict[str, Any]]:
+        archive_dp = resource_dps[0]
+        images_dp, split_dp = Demultiplexer(
+            archive_dp, 2, self._classify_archive, drop_none=True, buffer_size=INFINITE_BUFFER_SIZE
+        )
+        split_dp = Filter(split_dp, path_comparator("name", f"{config.split}.json"))
+        split_dp = JsonParser(split_dp)
+        split_dp = Mapper(split_dp, getitem(1))
+        split_dp: IterDataPipe[List[str]] = MappingIterator(split_dp, drop_key=True)
+        split_dp = UnBatcher(split_dp)
```

Comment:
While looking into the raw data again, I saw that there are `{train, test}.txt` files that contain all images for the given split. That seems a lot easier than what we are doing here. Was there a reason to use the JSON files in the first place?

Reply:
No particular reason except using the same JSON files as in the previous dataset. I guess using `LineReader` is what you have in mind?

Reply:
Yeah, if I'm not mistaken you can do

```suggestion
        split_dp = Filter(split_dp, path_comparator("name", f"{config.split}.json"))
        split_dp = LineReader(split_dp, decode=True, return_path=False)
```

which is a good deal simpler without losing any information.

Reply:
That's even better, thanks for spotting this simplification. :+1:

Reply:
I will update the mock accordingly.